packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts

```diff
@@ -0,0 +1,200 @@
+      dataSourceMetadata,
+      workspaceDataSource,
+      connectedAccount[0].id,
+    );
+
+    return threads;
+  }
+
+  async fetchWorkspaceMemberMessages(
+    workspaceId: string,
+    workspaceMemberId: string,
+    maxResults = 500,
+  ): Promise<any> {
+    const dataSourceMetadata =
+      await this.dataSourceService.getLastDataSourceMetadataFromWorkspaceIdOrFail(
+        workspaceId,
+      );
+
+    const workspaceDataSource = await this.typeORMService.connectToDataSource(
+      dataSourceMetadata,
+    );
+
+    const connectedAccount = await workspaceDataSource?.query(
+      `SELECT * FROM ${dataSourceMetadata.schema}."connectedAccount" WHERE "provider" = 'gmail' AND "accountOwnerId" = $1`,
+      [workspaceMemberId],
+    );
+
+    const accessToken = connectedAccount[0].accessToken;
+    const refreshToken = connectedAccount[0].refreshToken;
+
+    const gmail = await this.getGmailClient(refreshToken);
+
+    const messages = await gmail.users.messages.list({
+      userId: 'me',
+      maxResults,
+    });
+
+    const messagesData = messages.data.messages;
+
+    if (!messagesData) {
+      return;
+    }
+
+    const messageQueries = messagesData.map((message) => ({
+      uri: '/gmail/v1/users/me/messages/' + message.id,
+    }));
+
+    const messagesResponse =
+      await this.fetchBatchMessagesService.fetchAllByBatches(
+        messageQueries,
+        accessToken,
+      );
+
+    await this.saveMessages(
+      messagesResponse,
+      dataSourceMetadata,
+      workspaceDataSource,
+    );
+
+    return messages;
+  }
+
+  async getGmailClient(refreshToken) {
+    const gmailClientId = this.environmentService.getAuthGoogleClientId();
+
+    const gmailClientSecret =
+      this.environmentService.getAuthGoogleClientSecret();
+
+    const oAuth2Client = new google.auth.OAuth2(
+      gmailClientId,
+      gmailClientSecret,
+    );
+
+    oAuth2Client.setCredentials({
+      refresh_token: refreshToken,
+    });
+
+    return google.gmail({
+      version: 'v1',
+      auth: oAuth2Client,
+    });
+  }
+
+  async saveMessageThreads(
+    threads,
+    dataSourceMetadata,
+    workspaceDataSource,
+    connectedAccountId,
+  ) {
+    const messageChannel = await workspaceDataSource?.query(
+      `SELECT * FROM ${dataSourceMetadata.schema}."messageChannel" WHERE "connectedAccountId" = $1`,
+      [connectedAccountId],
+    );
+
+    for (const thread of threads) {
+      await workspaceDataSource?.query(
+        `INSERT INTO ${dataSourceMetadata.schema}."messageThread" ("externalId", "subject", "messageChannelId", "visibility") VALUES ($1, $2, $3, $4)`,
+        [thread.id, thread.snippet, messageChannel[0].id, 'default'],
+      );
+    }
+  }
+
+  async saveMessages(messages, dataSourceMetadata, workspaceDataSource) {
+    for (const message of messages) {
+      const {
+        externalId,
+        headerMessageId,
+        subject,
+        messageThreadId,
+        from,
+        body,
+      } = message;
```

Comment:
Not sure why we need to deconstruct here? Not a strong opinion about it, just curious!